src/counts.jl

```diff
@@ -238,12 +238,48 @@ If a weighting vector `wv` is specified, the sum of the weights is used rather t
 raw counts.
 """
 function addcounts!(cm::Dict{T}, x::AbstractArray{T}) where T
+    # if it's of bits type then can speed things up using radix sort
+    # from heuristics it was found that 2^16 = 65536 is the point at which
+    # radixsort is more performant
+    if isbits(T) & length(x) > 65536
```

Comment:
Radix sort is not safe for any `isbits` type. Any package is free to implement `isbits` types which arbitrary semantics for `==`. So we should hardcode a list of Base types known to be safe, via a function called e.g. `radix_sort_safe`.

Also, is the threshold the same for all types? Maybe not, and the function could be used to adapt that.

Reply:
Would doing that limit the number of types fpr which this will have a performancr boost? If they define isbits when the tyoe isn't bits i.e. can't be shifted by >> then they have done something "wrong" almost.

Reply:
No. Imagine a type for which `==` can be `true` even though the bits representation is not equal. For example, a modulus integer type could decide that `CustomModInt(1) == CustomModInt(10) == CustomModInt(20)`. If internally the value is stored as `1`, `10` and `20`, the sorting based on bytes won't be correct.

Reply:
So `radix_sort_safe` should return false and hence it will ignore this type? This means we have to define `radix_sort_safe` like this

```julia
radix_sort_safe(::Any) = false
radix_sort_safe(::T<:Number) = true
...etc
```
What if they define `T` to be like you mentioned but still define it as a subtype of Numbers? So we have to define radix_sort_saft for Float64, Int64 etc etc. But then it wouldn't be available for other legitimate types that are bits and didn't choose to overwrite `==`. In general seems to be a hard issue. So every bits type that wants to be able to benefit sort will have to know to define the radxisort_safe method for their own type? This seems to be very beginner unfriendly.

I feel that those that overload `==` are in the minority, so we shouldn't cater to them, I can define `radix_sort_safe` to be `isbits(T)` and let them (the minority) override `radix_sort_safe` would be a better solution?

thoughts?

Reply:
> So every bits type that wants to be able to benefit sort will have to know to define the radxisort_safe method for their own type? This seems to be very beginner unfriendly.

Yes, that's annoying, but only package authors will have to do it. OTC, allowing radix sort by default could give incorrect results to basic users who don't even know what radix sort is, which is much more problematic.

It would make sense to define that method in SortingAlgorithms so that it can be shared across packages. I'd argue that sorting using radix sort should throw an error by default in that package too, with an error indicating how to override this when appropriate. Would you file an issue there?